### PR TITLE
Story/jenkins 42791 scm source additions

### PR DESCRIPTION
**Pull request by cliffmeyers**
# Description
- This PR adds a `scmSource` property to multibranch pipelines and org folders so that the client can be aware of which scm provider it's tied to and the `apiUrl` so that the credential can be properly looked up and later created if not available.
- Below are some examples of the new responses..

GitHub Cloud
```
{
    "_class": "io.jenkins.blueocean.blueocean_github_pipeline.GithubOrganizationFolder",
    "_links": {},
    "actions": [],
    "displayName": "Cliff Meyers",
    "fullDisplayName": "Cliff%20Meyers",
    "fullName": "cliffmeyers",
    "name": "cliffmeyers",
    "organization": "jenkins",
    "parameters": null,
    "permissions": null,
    "icon": {},
    "numberOfFolders": 2,
    "numberOfPipelines": 0,
    "pipelineFolderNames": [],
    "scanAllRepos": false,
    "scmSource": {
        "_class": "io.jenkins.blueocean.rest.impl.pipeline.scm.ScmSourceImpl",
        "apiUrl": "https://api.github.com",
        "id": "github"
    },
    "latestRun": {},
    "repos": {}
}
```

GitHub Enterprise
```
{
    "_class": "io.jenkins.blueocean.blueocean_github_pipeline.GithubOrganizationFolder",
    "_links": {},
    "actions": [],
    "displayName": "Dev1",
    "fullDisplayName": "Dev1",
    "fullName": "dev1",
    "name": "dev1",
    "organization": "jenkins",
    "parameters": null,
    "permissions": null,
    "icon": {},
    "numberOfFolders": 2,
    "numberOfPipelines": 0,
    "pipelineFolderNames": [],
    "scanAllRepos": false,
    "scmSource": {
        "_class": "io.jenkins.blueocean.rest.impl.pipeline.scm.ScmSourceImpl",
        "apiUrl": "https://github.beescloud.com/api/v3",
        "id": "github-enterprise"
    },
    "latestRun": {},
    "repos": {}
}
```

Bitbucket Cloud
```
{
    "_class": "io.jenkins.blueocean.rest.impl.pipeline.MultiBranchPipelineImpl",
    "_links": {},
    "actions": [],
    "displayName": "jenkins-test-private-1",
    "fullDisplayName": "jenkins-test-private-1",
    "fullName": "jenkins-test-private-1",
    "name": "jenkins-test-private-1",
    "organization": "jenkins",
    "parameters": null,
    "permissions": {},
    "estimatedDurationInMillis": 1844,
    "numberOfFolders": 0,
    "numberOfPipelines": 2,
    "pipelineFolderNames": [],
    "weatherScore": 0,
    "branchNames": [],
    "numberOfFailingBranches": 2,
    "numberOfFailingPullRequests": 0,
    "numberOfSuccessfulBranches": 0,
    "numberOfSuccessfulPullRequests": 0,
    "scmSource": {
        "_class": "io.jenkins.blueocean.rest.impl.pipeline.scm.ScmSourceImpl",
        "apiUrl": "https://bitbucket.org",
        "id": "bitbucket-cloud"
    },
    "totalNumberOfBranches": 2,
    "totalNumberOfPullRequests": 0
}
```

Bitbucket Server
```
{
    "_class": "io.jenkins.blueocean.rest.impl.pipeline.MultiBranchPipelineImpl",
    "_links": {},
    "actions": [],
    "displayName": "repo-2",
    "fullDisplayName": "repo-2",
    "fullName": "repo-2",
    "name": "repo-2",
    "organization": "jenkins",
    "parameters": null,
    "permissions": {},
    "estimatedDurationInMillis": 293,
    "numberOfFolders": 0,
    "numberOfPipelines": 2,
    "pipelineFolderNames": [],
    "weatherScore": 100,
    "branchNames": [],
    "numberOfFailingBranches": 0,
    "numberOfFailingPullRequests": 0,
    "numberOfSuccessfulBranches": 2,
    "numberOfSuccessfulPullRequests": 0,
    "scmSource": {
        "_class": "io.jenkins.blueocean.rest.impl.pipeline.scm.ScmSourceImpl",
        "apiUrl": "http://localhost:7990/bitbucket",
        "id": "bitbucket-server"
    },
    "totalNumberOfBranches": 2,
    "totalNumberOfPullRequests": 0
}
```

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by cliffmeyers** [comment]
@vivek these are the SCM changes we discussed the other day. PTAL :)

File: blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/cloud/BitbucketCloudScm.java [hunk 1]
```diff
@@ -13,9 +13,9 @@
  * @author Vivek Pandey
  */
 public class BitbucketCloudScm extends AbstractBitbucketScm {
+    public static final String ID = "bitbucket-cloud";
+    public static final String API_URL = "https://bitbucket.org";
     static final String DOMAIN_NAME="blueocean-bitbucket-cloud-domain";
-    static final String ID = "bitbucket-cloud";
-    static final String API_URL = "https://bitbucket.org/api/2.0/";
```
Comment (cliffmeyers) [inline blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/cloud/BitbucketCloudScm.java]: This didn't appear to be used anywhere and the `serverUrl` returned from the Bitbucket Branch Source api was actually just the protocol+domain, so I changed it.

File: blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/AbstractBitbucketScmContentProvider.java [hunk 2]
```diff
@@ -133,13 +134,15 @@ public GitContent getContent() {
     }
 
     @SuppressWarnings("unchecked")
-    @Override
-    public boolean support(@Nonnull Item item) {
+    @CheckForNull
```
Comment (sophistifunk) [inline blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/AbstractBitbucketScmContentProvider.java]: What's `@CheckForNull` actually mean? I googled it and came away more confused than when I started, which makes me sad. Is it just marking the return value as nullable?

**Comment by michaelneale** [comment]
I am ok with this as long as @vivek is. 

@sophistifunk had a Q about CheckForNull - I guess that is a Jenkins ism?

**Comment by sophistifunk** [comment]
Nah it's in `javax.annotations` so it's not a Jenkinsism. But I'm just curious, it doesn't reflect on this code.

File: blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/AbstractBitbucketScmContentProvider.java (hunk 2 above)
Comment (michaelneale) [inline blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/AbstractBitbucketScmContentProvider.java]: it's a rule for findbugs apparently - to ensure nulls are defended against where it is used I guess.

**Comment by cliffmeyers** [comment]
cc @vivek as a reminder that a review is needed here when you're able. Thanks!

**Comment by vivek** [comment]
@cliffmeyers is this waiting for something? If its all good to merge, we should get it in as it can help us add better support for  bitbucket/MBP support in pipeline editor.

**Comment by cliffmeyers** [comment]
@vivek I'll see if I can get this merged tomorrow morning. I'm on and off planes for the next seven or so hours. :)

**Comment by vivek** [comment]
@cliffmeyers didn't mean to interrupt your vacation:) @michaelneale is going to look in to it and if everything works then we can merge it. Sounds good?

**Comment by michaelneale** [comment]
this works for me - so unless @cliffmeyers says more to do - will merge it tomorrow

**Comment by michaelneale** [comment]
@vivek you can merge this whenever you need it

**Comment by cliffmeyers** [comment]
@vivek not at all, I was just trying to be funny ;)

**Pull request merged**